File: python-gdb-activities-20/activity20-solution/gdb/ui/account_ui.py
# gdb/ui/account_ui.py
import sys
from gdb.service.account_service import AccountService
from gdb.exceptions.account_exception import AccountException

class AccountUI:
    """Interactive console interface for banking operations."""

    def __init__(self, service: AccountService) -> None:
        self._service = service

    def start(self) -> None:
        while True:
            self.display_main_menu()
            choice = self.read_int("Enter your choice: ")
            try:
                if choice == 1:
                    self.handle_open_account()
                elif choice == 2:
                    self.handle_deposit()
                elif choice == 3:
                    self.handle_withdraw()
                elif choice == 4:
                    self.handle_transfer()
                elif choice == 5:
                    self.handle_close_account()
                elif choice == 6:
                    self.handle_view_account()
                elif choice == 7:
                    self.handle_view_transactions()
                elif choice == 8:
                    print("Thank you! Goodbye.")
                    break
                else:
                    print("Invalid choice. Please enter 1-8.")
            except Exception as e:
                print(f"ERROR: {e}")

    def display_main_menu(self) -> None:
        print("\n" + "=" * 50)
        print("     GLOBAL DIGITAL BANK (GDB) — MAIN MENU")
        print("=" * 50)
        print("1. Open Account")
        print("2. Deposit Funds")
        print("3. Withdraw Funds")
        print("4. Transfer Funds")
        print("5. Close Account")
        print("6. View Account Details")
        print("7. View Transaction History")
        print("8. Exit")
        print("=" * 50)

    def handle_open_account(self) -> None:
        acc_type = self.read_string("Account Type (Savings/Current/FixedDeposit/Salary): ")
        name = self.read_string("Account Holder Name: ")
        age = self.read_int("Age: ")
        initial_balance = self.read_double("Initial Balance: ")
        tenure = self.read_int("Customer Tenure in Years (0 for new): ")
        acc = self._service.open_account(acc_type, name, age, initial_balance, tenure)
        pin = self.read_int("Set 4-digit PIN (1000-9999): ")
        acc.set_pin(pin)
        print(f"SUCCESS: Account #{acc.get_account_number()} created successfully!")
        print(acc.get_account_info())

    def handle_deposit(self) -> None:
        acc_no = self.read_int("Account Number: ")
        amount = self.read_double("Amount to deposit: ")
        txn = self._service.deposit(acc_no, amount)
        print(f"SUCCESS: {txn.get_receipt()}")

    def handle_withdraw(self) -> None:
        acc_no = self.read_int("Account Number: ")
        amount = self.read_double("Amount to withdraw: ")
        pin = self.read_int("4-digit PIN: ")
        txn = self._service.withdraw(acc_no, amount, pin)
        print(f"SUCCESS: {txn.get_receipt()}")

    def handle_transfer(self) -> None:
        from_acc = self.read_int("From Account Number: ")
        to_acc = self.read_int("To Account Number: ")
        amount = self.read_double("Amount to transfer: ")
        pin = self.read_int("Sender 4-digit PIN: ")
        txn = self._service.transfer(from_acc, to_acc, amount, pin)
        print(f"SUCCESS: {txn.get_receipt()}")

    def handle_close_account(self) -> None:
        acc_no = self.read_int("Account Number: ")
        pin = self.read_int("4-digit PIN: ")
        self._service.close_account(acc_no, pin)
        print(f"SUCCESS: Account #{acc_no} closed.")

    def handle_view_account(self) -> None:
        acc_no = self.read_int("Account Number: ")
        acc = self._service.get_account(acc_no)
        if acc is None:
            print(f"Account #{acc_no} not found.")
        else:
            print(acc.get_account_info())

    def handle_view_transactions(self) -> None:
        txns = self._service.get_transaction_history()
        if not txns:
            print("No transactions logged.")
            return
        print("\n--- TRANSACTION HISTORY LOG ---")
        for i, t in enumerate(txns, 1):
            print(f"[{i}] {t.get_receipt()}")

    def read_int(self, prompt: str) -> int:
        while True:
            try:
                val = input(prompt).strip()
                return int(val)
            except ValueError:
                print("Invalid integer. Please try again.")

    def read_double(self, prompt: str) -> float:
        while True:
            try:
                val = input(prompt).strip()
                return float(val)
            except ValueError:
                print("Invalid number. Please try again.")

    def read_string(self, prompt: str) -> str:
        while True:
            val = input(prompt).strip()
            if val:
                return val
            print("Input cannot be empty.")

File: java-gdb-activities-20/activity20-solution/src/com/gdb/ui/AccountUI.java
package com.gdb.ui;

import com.gdb.domain.IAccount;
import com.gdb.domain.Transaction;
import com.gdb.command.TransactionCommand;
import com.gdb.service.AccountService;
import com.gdb.exceptions.AccountException;

import java.util.*;

public class AccountUI {
    private final AccountService service;
    private final Scanner scanner;

    public AccountUI(AccountService service) {
        this.service = service;
        this.scanner = new Scanner(System.in);
    }

    public void start() {
        while (true) {
            displayMainMenu();
            int choice = readInt("Enter your choice: ");
            try {
                switch (choice) {
                    case 1: handleOpenAccount(); break;
                    case 2: handleDeposit(); break;
                    case 3: handleWithdraw(); break;
                    case 4: handleTransfer(); break;
                    case 5: handleCloseAccount(); break;
                    case 6: handleViewAccount(); break;
                    case 7: handleViewTransactions(); break;
                    case 8: System.out.println("Thank you! Goodbye."); return;
                    default: System.out.println("Invalid choice. Please enter 1-8.");
                }
            } catch (Exception e) {
                System.out.println("ERROR: " + e.getMessage());
            }
        }
    }

    private void displayMainMenu() {
        System.out.println("\n========================================");
        System.out.println("   GLOBAL DIGITAL BANK");
        System.out.println("========================================");
        System.out.println("1. Open Account");
        System.out.println("2. Deposit");
        System.out.println("3. Withdraw");
        System.out.println("4. Transfer");
        System.out.println("5. Close Account");
        System.out.println("6. View Account Details");
        System.out.println("7. View Transaction History");
        System.out.println("8. Exit");
        System.out.println("========================================");
    }

    private void handleOpenAccount() throws Exception {
        System.out.println("\n--- Open Account ---");
        String type = readString("Account Type (Savings/Current/FixedDeposit/Salary): ");
        String name = readString("Name: ");
        int age = readInt("Age: ");
        double amount = readDouble("Initial Balance: ");

        IAccount acc = service.openAccount(type, name, age, amount);
        System.out.println("SUCCESS: " + acc.getAccountInfo());

        int pin = readInt("Set 4-digit PIN: ");
        acc.setPin(pin);
        System.out.println("PIN set successfully.");
    }

    private void handleDeposit() throws Exception {
        System.out.println("\n--- Deposit ---");
        int accNo = readInt("Account Number: ");
        double amount = readDouble("Amount to deposit: ");
        Transaction txn = service.deposit(accNo, amount);
        System.out.println("SUCCESS: Deposited Rs. " + amount + " to #" + accNo + ". New balance: Rs. " + txn.getBalanceAfter());
    }

    private void handleWithdraw() throws Exception {
        System.out.println("\n--- Withdraw ---");
        int accNo = readInt("Account Number: ");
        double amount = readDouble("Amount to withdraw: ");
        int pin = readInt("PIN: ");
        Transaction txn = service.withdraw(accNo, amount, pin);
        System.out.println("SUCCESS: Withdrew Rs. " + amount + " from #" + accNo + ". New balance: Rs. " + txn.getBalanceAfter());
    }

    private void handleTransfer() throws Exception {
        System.out.println("\n--- Transfer ---");
        int fromAcc = readInt("From Account: ");
        int toAcc = readInt("To Account: ");
        double amount = readDouble("Amount: ");
        int pin = readInt("PIN: ");
        Transaction txn = service.transfer(fromAcc, toAcc, amount, pin);
        System.out.println("SUCCESS: Transferred Rs. " + amount + " from #" + fromAcc + " to #" + toAcc);
    }

    private void handleCloseAccount() throws Exception {
        System.out.println("\n--- Close Account ---");
        int accNo = readInt("Account Number: ");
        int pin = readInt("PIN: ");
        service.closeAccount(accNo, pin);
        System.out.println("SUCCESS: Account #" + accNo + " has been closed.");
    }

    private void handleViewAccount() {
        System.out.println("\n--- View Account Details ---");
        int accNo = readInt("Account Number: ");
        IAccount acc = service.getAccount(accNo);
        if (acc == null) {
            System.out.println("ERROR: Account not found: " + accNo);
        } else {
            System.out.println(acc.getAccountInfo());
        }
    }

    private void handleViewTransactions() {
        System.out.println("\n--- Transaction History ---");
        List<TransactionCommand> history = service.getTransactionHistory();
        if (history.isEmpty()) {
            System.out.println("No transactions logged.");
        } else {
            for (int i = 0; i < history.size(); i++) {
                System.out.println("  [" + (i + 1) + "] " + history.get(i).getTransaction());
            }
        }
    }

    private int readInt(String prompt) {
        while (true) {
            System.out.print(prompt);
            try {
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Invalid integer number. Please try again.");
            }
        }
    }

    private double readDouble(String prompt) {
        while (true) {
            System.out.print(prompt);
            try {
                return Double.parseDouble(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Invalid decimal number. Please try again.");
            }
        }
    }

    private String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }
}
